### PR TITLE
Changing object.id to object.pk and using `quote` to handle non-url-s…

### LINKED ISSUE #9205: wagtail.admin.views.generic.EditView uses object.id which is not present in every object.
### Issue Summary

As mentioned in https://github.com/wagtail/wagtail/pull/9204#discussion_r971114461, `wagtail.admin.views.generic.EditView` uses `object.id` in order to refer to an URL. However, not every object has the `id` field.

### Steps to Reproduce

1. Use `wagtail.admin.views.generic.EditView` instead of `wagtail.users.views.users.create`.
2. Runtests using `./runtests.py wagtail.users.tests.test_admin_views`.
3. As the testrunner reaches the `get_delete_url`, it throws the following error:
> AttributeError: 'CustomUser' object has no attribute 'id'

**Note: The same issue exists with `get_success_buttons` in addition to `get_delete_url`.**

### Technical details

-   Wagtail version: 4.0a0.

### PR DESCRIPTION
**Pull request by mehrdadmoradii**
fixes #9205

Changing the `args` argument to make use of `pk` instead of `id` as a primary key to retrieve an object. Also, ensure we're using the `quote`/`unquote` consistently to handle non-url-safe primary keys.

### CODE REVIEW

File: wagtail/admin/views/generic/models.py [hunk 1]
```diff
@@ -406,7 +406,7 @@ def get_success_message(self, instance):
     def get_success_buttons(self):
         return [
             messages.button(
-                reverse(self.edit_url_name, args=(self.object.id,)), _("Edit")
```
[the remark this answers is not in the document]
Comment (lb-) [reply wagtail/admin/views/generic/models.py:409]: sounds good - add this refined approach to the PR and we can take a look.

**Comment by lb-** [comment]
@mehrdadmoradii  been looking at this and the code change seems solid, however, this will need some testing.

It does not look like we have tests in place for these generic views used outside the main models, so here is a potential plan to get this tested.

* Create a new testapp model, check first as there may be one, that does not use an integer primary key - wagtail/test/testapp/models.py (in the `""" ... """` pydoc for the model, add a note that this is a model that intentionally does not use an integer key)
* Run testapp migrations for this model - https://docs.wagtail.org/en/stable/contributing/developing.html#running-migrations-for-the-test-app-models
* Set up some test data for this model (if needed) in fixtures wagtail/test/testapp/fixtures/test.json
* I think this file would be the most suitable place for these 'generic' model views - `wagtail/admin/tests/test_views.py`, alternatively you can create a new file, something like `wagtail/admin/tests/test_views_generic.py`.
* From here you will need to write a test for each of the generic view models, just start with one first and get that passing then build on it.

I know that writing these tests will probably take 10x as long as actually fixing the problem. But having these tests in place will be a huge critical improvement and ensure that we can continue to add better support for non-integer ids with reasonable confidence it will not break in the future.

**Comment by mehrdadmoradii** [comment]
@lb- got it. I am going to work on that.

**Comment by gasman** [comment]
@mehrdadmoradii Your [resolving the merge conflict](https://github.com/wagtail/wagtail/pull/9221/commits/b1d105a11bf5c17068d3d9c7abe1bf490d69ebc3) seems to have reverted a huge number of changes from the Wagtail main branch - please re-check.

**Review by lb- — COMMENTED** [review]
@mehrdadmoradii  I have rebased this - done some tests locally and it is looking good.
I have also removed the `get_success_buttons` overrides from the user views.

Will check on the CI and merge in if all good.

One other note - not sure why but your test app migration had a very old number, `8` but the latest was `73` so I have put the test app migration last & renamed accordingly. This is probably not critical but fixed it up nonetheless.

**Comment by gasman** [comment]
> One other note - not sure why but your test app migration had a very old number, `8` but the latest was `73` so I have put the test app migration last & renamed accordingly. This is probably not critical but fixed it up nonetheless.

@lb- Test app migrations were squashed in #8783, so the sequence started again from 0002.

**Comment by lb-** [comment]
Thanks @gasman I'll rename it back to 8 and fix that up before merging.

I still want to do a few more manual tests on this but hopefully will get it in tonight

**Pull request merged**